From be63e228451d053665acb004fd8a8e17a48faa71 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Sat, 23 Sep 2023 11:56:57 +0200
Subject: [PATCH 0113/1010] genirq: irq_pipeline: do not request IPI on
 uniprocessor

The base interrupt controller of a uniprocessor machine provides no
IPI. When running a SMP kernel over such machine, do not attempt to
request the oob call function IPI if the hardware cannot support
multiple CPUs.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 kernel/irq/pipeline.c | 18 +++++++++++-------
 1 file changed, 11 insertions(+), 7 deletions(-)

diff --git a/kernel/irq/pipeline.c b/kernel/irq/pipeline.c
index 89119db8da..2f3df68fe0 100644
--- a/kernel/irq/pipeline.c
+++ b/kernel/irq/pipeline.c
@@ -1838,14 +1838,18 @@ void __init irq_pipeline_init(void)
 
 #ifdef CONFIG_SMP
 	/*
-	 * Now that the arch-specific code has prepared for handling
-	 * the oob call function IPI, hook the latter to the generic
-	 * call queue flushing routine.
+	 * Hook the remote call IPI which smp_call_function_oob()
+	 * sends to notify remote CPUs of pending work.
+	 *
+	 * CAUTION: the base interrupt controller of a uniprocessor
+	 * machine provides no IPI: skip the operation if the hardware
+	 * cannot support multiple CPUs.
 	 */
-	if (__request_percpu_irq(CALL_FUNCTION_OOB_IPI,
-				 smp_call_function_ipi_handler,
-				 IRQF_OOB, "Out-of-band function call IPI",
-				 &pipeline_percpu_data))
+	if (num_possible_cpus() > 1 &&
+		__request_percpu_irq(CALL_FUNCTION_OOB_IPI,
+				smp_call_function_ipi_handler,
+				IRQF_OOB, "out-of-band function call IPI",
+				&pipeline_percpu_data))
 		WARN_ON(1);
 #endif
 
-- 
2.30.2

